src/metabase/sync/analyze/fingerprint.clj

```diff
@@ -43,24 +46,74 @@
   (sync-util/with-error-handling (format "Error generating fingerprint for %s" (sync-util/name-for-logging field))
     (when-let [fingerprint (fingerprint field)]
       (log/debug (format "Saving fingerprint for %s" (sync-util/name-for-logging field)))
+      ;; All Fields who get new fingerprints should get marked as having the latest fingerprint version, but we'll
+      ;; clear their values for `last_analyzed`. This way we know these fields haven't "completed" analysis for the
+      ;; latest fingerprints.
       (db/update! Field (u/get-id field)
-        :fingerprint fingerprint))))
+        :fingerprint         fingerprint
+        :fingerprint_version i/latest-fingerprint-version
+        :last_analyzed       nil))))
 
 
-;;; +------------------------------------------------------------------------------------------------------------------------+
-;;; |                                          FINGERPRINTING ALL FIELDS IN A TABLE                                          |
-;;; +------------------------------------------------------------------------------------------------------------------------+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                    WHICH FIELDS NEED UPDATED FINGERPRINTS?                                     |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+;; Logic for building the somewhat-complicated query we use to determine which Fields need new Fingerprints
+;;
+;; This ends up giving us a SQL query that looks something like:
+;;
+;; SELECT *
+;; FROM metabase_field
+;; WHERE active = true
+;;   AND preview_display = true
+;;   AND visibility_type <> 'retired'
+;;   AND table_id = 1
+;;   AND ((fingerprint_version < 1 AND
+;;         base_type IN ("type/Longitude", "type/Latitude", "type/Integer"))
+;;        OR
+;;        (fingerprint_version < 2 AND
+;;         base_type IN ("type/Text", "type/SerializedJSON")))
+
+(s/defn ^:private ^:always-validate base-types->descendants :- #{su/FieldTypeKeywordOrString}
+  "Given a set of BASE-TYPES return an expanded set that includes those base types as well as all of their
+   descendants. These types are converted to strings so HoneySQL doesn't confuse them for columns."
+  [base-types :- #{su/FieldType}]
+  (->> (for [base-type base-types]
+         (cons base-type (descendants base-type)))
+       (reduce set/union)
+       (map u/keyword->qualified-name)
+       set))
+
+(s/defn ^:private ^:always-validate honeysql-for-fields-that-need-fingerprint-updating :- {:where s/Any}
+  "Return appropriate WHERE clause for all the Fields whose Fingerprint needs to be re-calculated."
+  ([]
+   {:where [:and
+            [:= :active true]
+            [:= :preview_display true]
+            [:not= :visibility_type "retired"]
```

Comment:
my understanding was that `:preview_display` relates to being displayed in the list views or not, but it would be viewable in an object detail. So, eg a big ass JSON column would might be turned off in list mode, but we'd still want to identify it as a JSON blob.

If it's a direct copy, we can fix it on a separate PR